opengrok-indexer/src/main/java/org/opengrok/indexer/util/Progress.java

```diff
@@ -0,0 +1,144 @@
+
+    private final Object sync = new Object();
+
+    /**
+     * @param logger logger instance
+     * @param suffix string suffix to identify the operation
+     * @param totalCount total count
+     */
+    public Progress(Logger logger, String suffix, long totalCount) {
+        this.logger = logger;
+        this.suffix = suffix;
+        this.totalCount = totalCount;
+
+        // Assuming printProgress configuration setting cannot be changed on the fly.
+        if (totalCount > 0 && RuntimeEnvironment.getInstance().isPrintProgress()) {
+            // spawn a logger thread.
+            run = true;
+            loggerThread = new Thread(this::logLoop,
+                    "progress-thread-" + suffix.replaceAll(" ", "_"));
+            loggerThread.start();
+        }
+    }
+
+    // for testing
+    Thread getLoggerThread() {
+        return loggerThread;
+    }
+
+    /**
+     * Increment counter. The actual logging will be done eventually.
+     */
+    public void increment() {
+        this.currentCount.incrementAndGet();
+
+        if (loggerThread != null) {
+            // nag the thread.
+            synchronized (sync) {
+                sync.notify();
+            }
+        }
+    }
+
+    private void logLoop() {
+        long cachedCount = 0;
+
+        while (true) {
+            long currentCount = this.currentCount.get();
+            Level currentLevel;
+
+            if (currentCount <= 1 || currentCount % 100 == 0) {
+                currentLevel = Level.INFO;
+            } else if (currentCount % 50 == 0) {
+                currentLevel = Level.FINE;
+            } else if (currentCount % 10 == 0) {
+                currentLevel = Level.FINER;
+            } else {
+                currentLevel = Level.FINEST;
+            }
+
+            // Do not log if there was no progress.
+            if (logger.isLoggable(currentLevel) && (cachedCount < currentCount)) {
+                logger.log(currentLevel, "Progress: {0} ({1}%) for {2}",
+                        new Object[]{currentCount, currentCount * 100.0f /
+                                totalCount, suffix});
+            }
+
+            if (!run) {
+                return;
+            }
+
+            cachedCount = currentCount;
+
+            // wait for event
+            try {
+                synchronized (sync) {
+                    if (!run) {
+                        // Loop once more to do the final logging.
+                        continue;
+                    }
+                    sync.wait();
+                }
+            } catch (InterruptedException e) {
+                logger.log(Level.WARNING, "logger thread interrupted");
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        if (loggerThread == null) {
+            return;
+        }
+
+        try {
+            run = false;
+            synchronized (sync) {
+                sync.notify();
+            }
+            loggerThread.join(0);
```

Comment:
Is perhaps `join()` more understandable by not having to remember that 0 means "wait forever"?